File: 2024/day22/solution.py
#!/usr/bin/env python3
import sys
# from functools import cache
from itertools import product
import time

sys.setrecursionlimit(1_000_000)
DEBUG = 0
FILENAME = 'example.txt'
D4 = [(-1, 0), (0, 1), (1, 0), (0, -1)]

for arg in sys.argv[1:]:
  if arg == '-i': FILENAME = 'input.txt'
  elif arg == '-e': FILENAME = 'example.txt'
  elif arg.startswith('-e'): FILENAME = f'example{arg[-1]}.txt'
  elif arg == '-d': DEBUG = 1
  elif arg == '-d2': DEBUG = 2
  elif arg == '-d3': DEBUG = 3
  else: raise Exception(f'unknown {arg=}')

with open(FILENAME, 'r') as f:
  inputs = list(map(int, f.read().splitlines()))


def evolve(value: int, iterations: int) -> int:
  orig_value = value
  for i in range(iterations):
    value ^= value << 6  # xor mult 64
    value &= 16777215  # mod 16777216
    value ^= value >> 5  # xor idiv 32
    value &= 16777215  # mod 16777216
    value ^= value << 11  # xor mult 2048
    value &= 16777215  # mod 16777216
    if DEBUG > 0: print(f'evolve {orig_value} {i+1} {value}')
  return value


def part1():
  # evolve(inputs[0], 10)
  # for value in inputs:
  #   print(f'{value=} {evolve(value,2000)}')
  total=sum( \
    evolve(value,2000) \
    for value in inputs \
  )
  print(f'part 1: {total=}')


def part2():
  sequence_totals: dict[tuple[int, int, int, int], int] = {}
  for value in inputs:
    # build dict of first price per sequence (cache_key) for each input value
    sequence_total: dict[tuple[int, int, int, int], int] = {}
    prices = []
    for _ in range(2000):
      value ^= value << 6  # xor mult 64
      value &= 16777215  # mod 16777216
      value ^= value >> 5  # xor idiv 32
      value &= 16777215  # mod 16777216
      value ^= value << 11  # xor mult 2048
      value &= 16777215  # mod 16777216
      prices.append(value % 10)
    for i in range(1996):
      cache_key = (
        prices[i + 1] - prices[i],
        prices[i + 2] - prices[i + 1],
        prices[i + 3] - prices[i + 2],
        prices[i + 4] - prices[i + 3],
      )
      if cache_key in sequence_total: continue
      sequence_total[cache_key] = prices[i + 4]
    # sum per-input value sequence prices into sequence_totals
    for k, v in sequence_total.items():
      sequence_totals[k] = sequence_totals.get(k, 0) + v

  # then we can just sort it to get the best sequence and total
  best_sequnce, best_total = sorted(sequence_totals.items(), key=lambda x: x[1], reverse=True)[0]

  print(f'part 2: {best_sequnce=} {best_total=}')


part1()
part2()

File: 2024/day22/solution.go
//go:build ignore

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

type Sequence struct {
	a, b, c, d int
}

func main() {
	// d4:=[][]int{{-1, 0},{0, 1},{1, 0},{0, -1}}
	filename, debug := parseArgs()
	fmt.Printf("filename: %s; debug: %d\n", filename, debug)
	data := readData(filename)
	if debug > 0 {
		fmt.Println(data)
	}
	part1(data, debug)
	part2(data, debug)
}

func part2(data []int, debug uint8) {
	sequenceTotals := map[Sequence]int{}
	for _, code := range data {
		codeSequences := map[Sequence]int{}
		// preallocate the entire window to avoid excessive memory reallocations in the loop
		window := make([]int, 0, 2000)
		for i := 0; i < 2000; i++ {
			// price from the initial code
			price := code % 10
			// window has an initial size of 0 and max size of 2000, so appends start at 0
			window = append(window, price)
			if i >= 4 {
				// based on the loop index
				sequence := Sequence{
					window[i-3] - window[i-4],
					window[i-2] - window[i-3],
					window[i-1] - window[i-2],
					window[i] - window[i-1],
				}
				// only add to code sequences if not exists
				if _, ok := codeSequences[sequence]; !ok {
					codeSequences[sequence] = price
				}
			}
			// the last evolve call is discarded but it's probably fine
			code = evolve(code)
		}
		if debug > 1 {
			fmt.Println("codeSequences", codeSequences)
		}
		// push the per-code values to the totals
		for key, val := range codeSequences {
			sequenceTotals[key] += val
		}
	}
	if debug > 0 {
		fmt.Println("sequenceTotals", sequenceTotals)
	}
	//now just select the highest value from sequenceTotals values
	highestPrice := 0
	var highestSequnce Sequence
	for sequence, price := range sequenceTotals {
		if price > highestPrice {
			highestPrice = price
			highestSequnce = sequence
		}
	}
	fmt.Println("part 2:", highestPrice, "sequence:", highestSequnce)
}

func part1(data []int, debug uint8) {
	total := uint64(0)
	for _, code := range data {
		for i := 0; i < 2000; i++ {
			code = evolve(code)
		}
		if debug > 0 {
			fmt.Println(code)
		}
		total += uint64(code)
	}
	fmt.Println("part 1:", total)
}

func evolve(input int) int {
	value := input
	value ^= (value << 6)
	value &= 16777215
	value ^= (value >> 5)
	value &= 16777215
	value ^= (value << 11)
	value &= 16777215
	return value
}

func parseArgs() (filename string, debug uint8) {
	filename = "example.txt"
	debug = 0
	for _, arg := range os.Args[1:] {
		switch arg {
		case "-i":
			filename = "input.txt"
		case "-e":
			filename = "example.txt"
		case "-e2":
			filename = "example2.txt"
		case "-e3":
			filename = "example3.txt"
		case "-e4":
			filename = "example4.txt"
		case "-d":
			debug = 1
		case "-d2":
			debug = 2
		case "-d3":
			debug = 3
		default:
			fmt.Printf("unknown arg: %s\n", arg)
		}
	}
	return
}

func readData(filename string) (data []int) {
	f, _ := os.Open(filename)
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		value, _ := strconv.Atoi(scanner.Text())
		data = append(data, value)
	}
	return data
}
